File: src/ados/services/ui/pages/details/about.py
"""About page — read-only system identity drilldown.

Shipped from the settings list as the bottom-most row. The body shows
agent version, board name + manufacturer, device id, MAC addresses
for the active network interfaces, build date (release-time stamp),
repo URL, and license. All values are read-only; the operator can
back out via the chevron.

Data sources:

* ``GET /api/v1/setup/status`` — top-level snapshot. Carries
  ``version``, ``device_id``, ``device_name``, board info via the
  setup state machine.
* ``/sys/class/net/<iface>/address`` — MAC addresses for any
  interface that exposes a sysfs file (eth0, wlan0, wlan1).
* ``/etc/ados/build.txt`` — optional release-time stamp written by
  the install script.

Each fetch is best-effort; missing values render as ``--``.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any, ClassVar

from PIL import Image, ImageDraw

from ados.services.ui.dashboards.components import primitives as p
from ados.services.ui.touch.events import TouchGesture

from ..base import HitZone, PageContext
from ._common import HEADER_H, draw_header_band

PAGE_W = 480
PAGE_H = 244


def _safe_get(d: Any, key: str, default: str = "--") -> str:
    if isinstance(d, dict):
        v = d.get(key)
        if v is not None and v != "":
            return str(v)
    return default


def _read_mac(iface: str) -> str:
    try:
        return Path(f"/sys/class/net/{iface}/address").read_text().strip() or "--"
    except OSError:
        return "--"


def _read_build_stamp() -> str:
    try:
        return Path("/etc/ados/build.txt").read_text().strip() or "--"
    except OSError:
        return "--"


def _read_board_name() -> str:
    try:
        from ados.hal.detect import detect_board

        board = detect_board()
        return getattr(board, "name", "") or "--"
    except Exception:
        return "--"


class AboutPage:
    """Read-only About modal pushed from the settings list."""

    id: ClassVar[str] = "details.about"
    refresh_hz: ClassVar[float] = 1.0

    def __init__(self) -> None:
        self._status: dict[str, Any] = {}

    async def on_enter(self, ctx: PageContext) -> None:
        ctx.logger.info("details_about_enter")
        await self._refresh(ctx)

    async def on_leave(self, ctx: PageContext) -> None:
        ctx.logger.info("details_about_leave")

    async def _refresh(self, ctx: PageContext) -> None:
        client = ctx.http
        if client is None:
            return
        try:
            r = await client.get("/api/v1/setup/status", timeout=1.5)
            if r.status_code == 200:
                blob = r.json() if callable(getattr(r, "json", None)) else {}
                if isinstance(blob, dict):
                    self._status = blob
        except Exception as exc:  # noqa: BLE001
            ctx.logger.debug("details_about_status_fetch_failed", error=str(exc))

    async def render(self, ctx: PageContext) -> Image.Image:
        palette = ctx.palette
        img = Image.new("RGB", (PAGE_W, PAGE_H), palette.bg_primary)
        draw_header_band(img, palette, "About")
        d = ImageDraw.Draw(img)

        version = _safe_get(self._status, "version")
        device_id = _safe_get(self._status, "device_id")
        device_name = _safe_get(self._status, "device_name")
        board = _read_board_name()
        eth_mac = _read_mac("eth0")
        wlan_mac = _read_mac("wlan0")
        build = _read_build_stamp()

        rows = [
            ("Agent", version),
            ("Board", board),
            ("Device ID", device_id),
            ("Device", device_name),
            ("eth0", eth_mac),
            ("wlan0", wlan_mac),
            ("Build", build),
            ("License", "GPLv3"),
            ("Repo", "github.com/altnautica/ADOSDroneAgent"),
        ]

        label_font = p.font("sans_bold", 11)
        value_font = p.font("mono_regular", 12)
        row_h = 22
        cy = HEADER_H + 6
        for label, value in rows:
            d.text(
                (16, cy),
                label.upper(),
                fill=palette.text_tertiary,
                font=label_font,
            )
            d.text(
                (140, cy - 1),
                value,
                fill=palette.text_primary,
                font=value_font,
            )
            cy += row_h
        return img

    def hit_zones(self, ctx: PageContext) -> list[HitZone]:
        return [HitZone(id="details.back", x=8, y=8, w=40, h=32)]

    async def on_touch(
        self,
        ctx: PageContext,
        zone: HitZone,
        gesture: TouchGesture,
    ) -> None:
        if zone.id == "details.back" and gesture.kind == "tap":
            await ctx.navigator.pop_modal(ctx=ctx)

File: crates/ados-display/src/pages/about.rs
//! About page — read-only system identity drilldown.
//!
//! Shown from the settings list as the bottom-most row. The body lists the
//! agent version, board name, device id, device name, the eth0 / wlan0 MAC
//! addresses, a release-time build stamp, the license, and the repo URL. Every
//! value is read-only; the operator backs out via the header chevron.
//!
//! The label/value rows are a fixed table: an uppercase caps label in the left
//! column (Sans Bold 11) and a mono value in the right column (Mono Regular 12),
//! 22 px per row. Missing identity fields render as `--`.
//!
//! All values come from [`PageContext::device`], which the render loop fills
//! from the setup-status snapshot and the HAL board detect plus the sysfs MAC
//! and build-stamp reads.

use crate::graphics::fonts::{FontFace, LoadedFont};
use crate::graphics::palette::Palette;
use crate::graphics::primitives::{text, Canvas};
use crate::pages::{blank_panel, HitZone, Page, PageContext};
use crate::widgets::{detail_back_zone, draw_detail_header, DETAIL_HEADER_H};

/// Header band height shared by every detail modal.
const HEADER_H: i32 = DETAIL_HEADER_H;
/// Vertical step between identity rows.
const ROW_H: i32 = 22;
/// Left edge of the caps labels.
const LABEL_X: i32 = 16;
/// Left edge of the mono values.
const VALUE_X: i32 = 140;
/// Repo URL shown on the last row.
const REPO_URL: &str = "github.com/altnautica/ADOSDroneAgent";

/// Resolve an optional identity field to its display text, defaulting to `--`
/// for an absent or empty value.
fn field(value: &Option<String>) -> String {
    match value {
        Some(s) if !s.is_empty() => s.clone(),
        _ => "--".to_string(),
    }
}

/// The about detail view, registered as `details.about`.
pub struct AboutDetailPage;

impl Page for AboutDetailPage {
    fn id(&self) -> &'static str {
        "details.about"
    }

    fn refresh_hz(&self) -> f32 {
        1.0
    }

    fn render(&self, ctx: &PageContext, palette: &Palette) -> Canvas {
        let mut canvas = blank_panel(palette);
        draw_detail_header(&mut canvas, palette, "About");

        let device = &ctx.device;
        let rows: [(&str, String); 9] = [
            ("Agent", field(&device.version)),
            ("Board", field(&device.board_name)),
            ("Device ID", field(&device.device_id)),
            ("Device", field(&device.device_name)),
            ("eth0", field(&device.mac_eth0)),
            ("wlan0", field(&device.mac_wlan0)),
            ("Build", field(&device.build_stamp)),
            ("License", "GPLv3".to_string()),
            ("Repo", REPO_URL.to_string()),
        ];

        let label_font = LoadedFont::new(FontFace::SansBold, 11);
        let value_font = LoadedFont::new(FontFace::MonoRegular, 12);
        let mut cy = HEADER_H + 6;
        for (label, value) in &rows {
            text(
                &mut canvas,
                &label_font,
                &label.to_ascii_uppercase(),
                LABEL_X,
                cy,
                palette.text_tertiary,
            );
            text(
                &mut canvas,
                &value_font,
                value,
                VALUE_X,
                cy - 1,
                palette.text_primary,
            );
            cy += ROW_H;
        }
        canvas
    }

    fn hit_zones(&self, _ctx: &PageContext) -> Vec<HitZone> {
        vec![detail_back_zone()]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::graphics::palette::DARK;
    use crate::pages::{HitAction, PANEL_W};

    #[test]
    fn about_renders_with_back_zone() {
        let page = AboutDetailPage;
        let ctx = PageContext::default();
        let c = page.render(&ctx, &DARK);
        assert_eq!(c.width(), PANEL_W);
        let zones = page.hit_zones(&ctx);
        assert_eq!(zones.len(), 1);
        assert_eq!(zones[0].action, HitAction::Back);
    }

    #[test]
    fn about_renders_identity_rows_with_real_values() {
        let page = AboutDetailPage;
        let mut ctx = PageContext::default();
        ctx.device.version = Some("0.49.39".to_string());
        ctx.device.board_name = Some("Raspberry Pi 4B".to_string());
        ctx.device.device_id = Some("ados-58c27faf".to_string());
        ctx.device.device_name = Some("groundnode".to_string());
        ctx.device.mac_eth0 = Some("dc:a6:32:00:11:22".to_string());
        ctx.device.mac_wlan0 = Some("dc:a6:32:00:11:23".to_string());
        ctx.device.build_stamp = Some("2026-06-01".to_string());
        // A populated context paints body content, so the body region is no
        // longer entirely the background color.
        let c = page.render(&ctx, &DARK);
        let mut body_inked = false;
        for y in (HEADER_H + 1)..280 {
            for x in 0..PANEL_W as i32 {
                if c.pixel(x, y) != DARK.bg_primary {
                    body_inked = true;
                    break;
                }
            }
            if body_inked {
                break;
            }
        }
        assert!(body_inked);
    }

    #[test]
    fn missing_fields_render_dashes() {
        assert_eq!(field(&None), "--");
        assert_eq!(field(&Some(String::new())), "--");
        assert_eq!(field(&Some("0.1".to_string())), "0.1");
    }
}
